route-rs-runtime/src/link/join_link.rs

```diff
@@ -1,51 +1,75 @@
 use crate::link::task_park::*;
-use crate::link::PacketStream;
+use crate::link::{Link, LinkBuilder, PacketStream, TokioRunnable};
 use crossbeam::atomic::AtomicCell;
 use crossbeam::crossbeam_channel;
 use crossbeam::crossbeam_channel::{Receiver, Sender};
 use futures::{task, Async, Future, Poll, Stream};
 use std::sync::Arc;
 
-pub struct JoinLink<Packet: Sized> {
-    pub ingressors: Vec<JoinIngressor<Packet>>,
-    pub egressor: JoinEgressor<Packet>,
+#[derive(Default)]
+pub struct JoinLinkBuilder<Packet: Sized + Send> {
+    in_streams: Option<Vec<PacketStream<Packet>>>,
+    queue_capacity: usize,
 }
 
-impl<Packet: Sized> JoinLink<Packet> {
-    pub fn new(input_streams: Vec<PacketStream<Packet>>, queue_capacity: usize) -> Self {
-        assert!(
-            input_streams.len() <= 1000,
-            format!("input_streams len {} > 1000", input_streams.len())
-        ); //Let's be reasonable here
+impl<Packet: Sized + Send> JoinLinkBuilder<Packet> {
+    pub fn new() -> Self {
+        JoinLinkBuilder {
+            in_streams: None,
+            queue_capacity: 10,
+        }
+    }
+
+    /// Changes queue_capacity, default value is 10.
+    /// Valid range is 1..=1000
+    pub fn queue_capacity(self, queue_capacity: usize) -> Self {
         assert!(
             queue_capacity <= 1000,
-            format!("Split Element queue_capacity: {} > 1000", queue_capacity)
+            format!("queue_capacity: {} > 1000", queue_capacity)
         );
         assert_ne!(queue_capacity, 0, "queue capacity must be non-zero");
 
-        let number_ingressors = input_streams.len();
-
-        let mut ingressors: Vec<JoinIngressor<Packet>> = Vec::new();
-        let mut from_ingressors: Vec<Receiver<Option<Packet>>> = Vec::new();
-        let mut task_parks: Vec<Arc<AtomicCell<TaskParkState>>> = Vec::new();
-
-        for input_stream in input_streams {
-            let (to_egressor, from_ingressor) =
-                crossbeam_channel::bounded::<Option<Packet>>(queue_capacity);
-            let task_park = Arc::new(AtomicCell::new(TaskParkState::Empty));
+        JoinLinkBuilder {
+            in_streams: self.in_streams,
+            queue_capacity,
+        }
+    }
+}
 
-            let ingressor =
-                JoinIngressor::new(input_stream, to_egressor.clone(), Arc::clone(&task_park));
-            ingressors.push(ingressor);
-            from_ingressors.push(from_ingressor);
-            task_parks.push(task_park);
+impl<Packet: Sized + Send + 'static> LinkBuilder<Packet, Packet> for JoinLinkBuilder<Packet> {
+    fn ingressors(self, in_streams: Vec<PacketStream<Packet>>) -> Self {
+        assert_ne!(in_streams.len(), 0, "Input stream vector can not be zero!");
```

Comment:
Should we assert a rough upper bound on the number of ingress streams for sanity? I can't imagine a good design having more than 20 or so.

Reply:
I set this to 1000, decided to let people be dumb if they want. Left it at 1000 for ingress streams, egressors, and queue length. I doubt anybody wants numbers this high for anything else.